src/esp/physics/bullet/BulletArticulatedObject.cpp

```diff
@@ -0,0 +1,808 @@
+        visualMeshInfo.filepath = visual.m_geometry.m_meshFileName;
+        // scale by radius as suggested by magnum docs
+        visualGeomComponent.scale(
+            Mn::Vector3(visual.m_geometry.m_capsuleRadius));
+        // Magnum capsule is Y up, URDF is Z up
+        visualGeomComponent.setTransformation(
+            visualGeomComponent.transformation() *
+            Mn::Matrix4::rotationX(Mn::Rad(M_PI_2)));
+        break;
+      case io::URDF::GEOM_CYLINDER:
+        visualMeshInfo.type = esp::assets::AssetType::PRIMITIVE;
+        // the default created primitive handle for the cylinder with radius 1
+        // and length 2
+        visualMeshInfo.filepath =
+            "cylinderSolid_rings_1_segments_12_halfLen_1_useTexCoords_false_"
+            "useTangents_false_capEnds_true";
+        visualGeomComponent.scale(
+            Mn::Vector3(visual.m_geometry.m_capsuleRadius,
+                        visual.m_geometry.m_capsuleHeight / 2.0,
+                        visual.m_geometry.m_capsuleRadius));
+        // Magnum cylinder is Y up, URDF is Z up
+        visualGeomComponent.setTransformation(
+            visualGeomComponent.transformation() *
+            Mn::Matrix4::rotationX(Mn::Rad(M_PI_2)));
+        break;
+      case io::URDF::GEOM_BOX:
+        visualMeshInfo.type = esp::assets::AssetType::PRIMITIVE;
+        visualMeshInfo.filepath = "cubeSolid";
+        visualGeomComponent.scale(visual.m_geometry.m_boxSize * 0.5);
+        break;
+      case io::URDF::GEOM_SPHERE: {
+        visualMeshInfo.type = esp::assets::AssetType::PRIMITIVE;
+        // default sphere prim is already constructed w/ radius 1
+        visualMeshInfo.filepath = "icosphereSolid_subdivs_1";
+        visualGeomComponent.scale(
+            Mn::Vector3(visual.m_geometry.m_sphereRadius));
+      } break;
+      case io::URDF::GEOM_MESH: {
+        visualGeomComponent.scale(visual.m_geometry.m_meshScale);
+        visualMeshInfo.filepath = visual.m_geometry.m_meshFileName;
+      } break;
+      case io::URDF::GEOM_PLANE:
+        Corrade::Utility::Debug()
+            << "Trying to add visual plane, not implemented";
+        // TODO:
+        visualSetupSuccess = false;
+        break;
+      default:
+        Corrade::Utility::Debug() << "BulletArticulatedObject::attachGeometry "
+                                     ": Unsupported visual type.";
+        visualSetupSuccess = false;
+        break;
+    }
+
+    // add the visual shape to the SceneGraph
+    if (visualSetupSuccess) {
+      assets::RenderAssetInstanceCreationInfo::Flags flags;
+      flags |= assets::RenderAssetInstanceCreationInfo::Flag::IsRGBD;
+      flags |= assets::RenderAssetInstanceCreationInfo::Flag::IsSemantic;
+      assets::RenderAssetInstanceCreationInfo creation(
+          visualMeshInfo.filepath, Mn::Vector3{1}, flags, DEFAULT_LIGHTING_KEY);
+
+      geomSuccess = const_cast<esp::assets::ResourceManager&>(resMgr_)
```

Comment:
Why const_cast? Is this a hack? Can we fix it?

Reply:
I refactored the pipeline to elevate this task to PhysicsManager which has non-const permissions on RM.